### PR TITLE
Strip `rust-` and `-rs` affixes in `cargo new`

### LINKED ISSUE #1532: When the user invokes `cargo new rust-foo`, name the resulting package `foo` instead
Every so often I spot packages with `rust-` or `-rs` in their name, e.g.

* https://crates.io/crates/rust-base58
* https://crates.io/crates/coroutine-rs (the maintainer renamed this one, thanks!)

These extra bits are redundant: the fact that it's on crates.io should tell us it's Rust already. So I think we should discourage names like this.

**Strawman proposal:** If the user invokes `cargo new` with a name of the form `rust-foo` or `foo-rs`, and the `--bin` option is *not* present, then the resulting `Cargo.toml` will have a `name` field of just `foo` instead.

I'm not sure if this'll require an RFC, but it's worth looking into as we get more users post-1.0.

### PR DESCRIPTION
**Pull request by lfairy**
If the user invokes `cargo new` with a name of the form `rust-foo` or `foo-rs` (or any variation of the two), and the `--bin` option is *not* present, then the resulting `Cargo.toml` will have a `name` field of just `foo` instead.

Closes #1532

### CODE REVIEW

**Comment by wycats** [comment]
At minimum we should explain to the user what has happened and provide a way to force it if the user really wants.

**Comment by tamird** [comment]
This should probably include a tests as well.

File: src/cargo/ops/cargo_new.rs [hunk 1]
```diff
@@ -56,12 +56,33 @@ pub fn new(opts: NewOptions, config: &Config) -> CargoResult<()> {
         return Err(human(&format!("Invalid character `{}` in crate name: `{}`",
                                   c, name)));
     }
-    mk(config, &path, name, &opts).chain_error(|| {
+    let name =
+        if opts.bin {
+            name.to_string()
```
Comment (tamird) [inline src/cargo/ops/cargo_new.rs]: all this string copying is not necessary, you ought to be able to continue to use the same underlying buffer

Comment (tamird) [inline src/cargo/ops/cargo_new.rs]: https://doc.rust-lang.org/nightly/std/primitive.str.html#method.trim_left_matches

**Comment by lfairy** [comment]
@tamird I added some tests, and eliminated the copying. I chose not to use `.trim_left_matches()` because it matches multiple times, which doesn't make sense here.

@wycats

Would you be happy with the following:

1. When this renaming happens, show this warning:

        Note: package will be named `foo`; use --name to override

2. Add a `--name` option for specifying an explicit package name

I haven't implemented this yet; just want to run it past you first.

File: src/cargo/ops/cargo_new.rs [hunk 2]
```diff
@@ -182,3 +202,20 @@ fn global_config(config: &Config) -> CargoResult<CargoNewConfig> {
...
+#[cfg(test)]
+mod tests {
+    use super::strip_rust_affixes;
+
+    #[test]
+    fn affixes_stripped() {
+        assert_eq!(strip_rust_affixes("rust-foo"), "foo");
+        assert_eq!(strip_rust_affixes("foo-rs"), "foo");
+        assert_eq!(strip_rust_affixes("rs_foo"), "foo");
+        // Only one affix is stripped
```
Comment (tamird) [inline src/cargo/ops/cargo_new.rs]: perhaps it shouldn't strip at all if there are multiple matches in this way? it's not clear to me which one would be correct to strip

**Comment by wycats** [comment]
@lfairy yes, that looks good.

**Comment by alexcrichton** [comment]
Could you add an integration test as well for this? Namely that the folder created is still called `foo-rs` but when running `cargo build` you get a library named `foo`

File: src/cargo/ops/cargo_new.rs (hunk 2 above)
Comment (lfairy) [inline src/cargo/ops/cargo_new.rs]: I wrote an alternate version that performs that check, but the code ended up too hard to understand.

In practice it shouldn't matter anyway. I scanned crates.io a few days ago and there are zero packages that have more than one `rust`/`rs` affix.

**Comment by lfairy** [comment]
Okay I've addressed @wycats and @alexcrichton's comments, and the tests do pass on my machine.

However the [Travis build](https://travis-ci.org/rust-lang/cargo/jobs/59830426) keeps failing in strange ways. Does anyone know why this is happening?

File: src/cargo/ops/cargo_new.rs [hunk 3]
```diff
@@ -46,10 +49,27 @@ pub fn new(opts: NewOptions, config: &Config) -> CargoResult<()> {
...
+                human(&format!("cannot create a project with a non-unicode name: {:?}",
+                               path.file_name().unwrap()))
+            }));
+            if opts.bin {
+                dir_name
+            } else {
+                let new_name = strip_rust_affixes(dir_name);
+                if new_name != dir_name {
+                    let message = format!(
+                        "Note: package will be named `{}`; use --name to override",
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_new.rs]: Conventionally our error messages start with lowercase letters I believe

**Comment by alexcrichton** [comment]
This looks good to me, thanks @lfairy! Our travis builds often can't handle the parallelism of the test suite (as it is a little resource intensive), leading to threads failing to spawn, leading to the test suite possibly hanging. Overall though it's just a false negative, so no need to worry!

File: src/cargo/ops/cargo_new.rs (hunk 3 above)
Comment (lfairy) [inline src/cargo/ops/cargo_new.rs]: The rest of the file seems to be inconsistent on the casing, so I wasn't sure about that.

I'll change it anyway though.

**Comment by lfairy** [comment]
@alexcrichton That makes sense, thanks!

**Comment by alexcrichton** [comment]
@bors: r+ 10fa4a128afd185ebc2f68b1b29832bd8df7edeb

Thanks!

**Comment by lfairy** [comment]
@alexcrichton The test failure should be fixed now.

**Pull request merged**